Christmas Edition - Audio_service addon to flet_video

syleishere opened the pull request:
## Description
Christmas Edition, full background play on android/IOS now, full bluetooth support, beautiful configurable notification screen, a function to update flet slider with percentages for position.

## Test Code

Nothing changes to normal example except more configurable options.

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] I signed the CLA.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally with my changes
- [ ] I have made corresponding changes to the [documentation](https://github.com/flet-dev/website) (if applicable)

## Additional details
Requires heavy mods to flet-build-template, I will submit a PR to -dev branch there like I have done with client dev area manifest here.

FeodorFitsner commented:
I love your mood here! 😅🥂

ndonkoHenri commented on sdk/python/packages/flet/src/flet/core/video.py:
```diff
@@ -96,6 +96,7 @@ def __init__(
         on_error: OptionalControlEventCallable = None,
         on_completed: OptionalControlEventCallable = None,
         on_track_changed: OptionalControlEventCallable = None,
+        on_percent_changed: OptionalControlEventCallable = None,
```
"percent" seems quite ambiguous here

syleishere replied:
I originally had it as on_pos_changed, but that seemed ambiguous to what it was doing, so I named it on_percent_changed to not confuse it with giving durations but giving percentages for the flet slider.

ndonkoHenri replied:
Actually, why send position values as percentage instead? Why not send "number of seconds" directly for example? Those who want percentage-conversion could then do it freely on python end. 

We could then name this event `on_position_change`, as the goal here is to share details on the new position.
What do you think?

(will appreciate test code to better review your PR)

syleishere replied:
When I looked at my slider config for displaying position changes in flet code went like:
```
        self.slider1 = ft.Slider(
            value=0,
            min=0,
            max=100,
            width=100,
            tooltip="Seek",
            divisions=100,
            label="{value}%",
            on_change=self.seek,
            thumb_color=ft.Colors.PRIMARY_CONTAINER,
            disabled=False,
            active_color=ft.Colors.ON_PRIMARY_CONTAINER,
            inactive_color=ft.Colors.ON_PRIMARY,
            on_focus=self.slider_focus,
            on_blur=self.slider_blur,
            on_change_end=self.on_change_end,
        )
```

I have 100 positions it can update with from 0-100 which seemed more than adequate, the issue is ft.Slider wants it in this format so then I was forced to do something like this in python:

```
                    duration = await self.config.video.get_duration_async(
                        wait_timeout=5
                    )
                    pos = await self.config.video.get_current_position_async(
                        wait_timeout=5
                    )
                    percent = int(pos / duration * 100)
                    print(f"Position changed to {pos} and duration is: {duration} Percent is: {percent}")
                    await self.config.video_lock.acquire()  # so we don't update slider while they're selecting a seek
                    self.slider1.value = percent
```

This seemed incredibly inefficient as I was using python to do the conversions, when we could use a faster language like dart to do the conversion for us then just send it to us.

syleishere replied:
I could create another handler for durations if that is what people want, but I think this is most efficient of working with ft.Slider.

syleishere replied:
I see what your saying though, looking at a music app from playstore, it keeps displaying time its been playing for the track.

syleishere replied:
What if passed duration, position and percentage, then everyone wins and they can do what they want with the values since they will always have all those in their flet program stored as variables just by enabling the handler? Also make the throttlespeed a configurable in milliseconds somewhere else.

ndonkoHenri replied:
> duration, position, percent

Sure, we can do it that way. Will you work on it? (Inspiration: [py](https://github.com/flet-dev/flet/blob/b842dc338ce128d5e328f98de615aa20d3b7c561/sdk/python/packages/flet/src/flet/core/dismissible.py#L299-L306), [dart](https://github.com/flet-dev/flet/blob/ab58bf9944fd6f2f5e69aae2417d31f2125b8f3a/packages/flet/lib/src/controls/interactive_viewer.dart#L134-L143))

ndonkoHenri commented:
Hey @syleishere, thanks for contributing. 
Will you mind providing code for review?

syleishere commented:
Sure, just adding on to example on Flet video page:

```
import random
import flet as ft


def main(page: ft.Page):
    page.theme_mode = ft.ThemeMode.LIGHT
    page.title = "TheEthicalVideo"
    page.window.always_on_top = True
    page.spacing = 20
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER

    def handle_pause(e):
        video.pause()
        print("Video.pause()")

    def handle_play_or_pause(e):
        video.play_or_pause()
        print("Video.play_or_pause()")

    def handle_play(e):
        video.play()
        print("Video.play()")

    def handle_stop(e):
        video.stop()
        print("Video.stop()")

    def handle_next(e):
        video.next()
        print("Video.next()")

    def handle_previous(e):
        video.previous()
        print("Video.previous()")

    def handle_volume_change(e):
        video.volume = e.control.value
        page.update()
        print(f"Video.volume = {e.control.value}")

    def handle_playback_rate_change(e):
        video.playback_rate = e.control.value
        page.update()
        print(f"Video.playback_rate = {e.control.value}")

    def handle_seek(e):
        video.seek(10000)
        print(f"Video.seek(10000)")

    def handle_add_media(e):
        video.playlist_add(random.choice(sample_media))
        print(f"Video.playlist_add(random.choice(sample_media))")

    def handle_remove_media(e):
        r = random.randint(0, len(video.playlist) - 1)
        video.playlist_remove(r)
        print(f"Popped Item at index: {r} (position {r+1})")

    def handle_jump(e):
        print(f"Video.jump_to(0)")
        video.jump_to(0)

    def handle_track_changed(e):
        index = int(e.data)
        print(f"Track changed to index: {index}, Currently Playing {SONG_LIST[index]}")

    def handle_position_changed(e):
        print(f"Position is {e.position} seconds, duration is {e.duration} seconds, percent done is {e.percent}%")

    def handle_error(e):
        print(f"Video error: {e}")

    SONG_LIST = [
        "https://user-images.githubusercontent.com/28951144/229373720-14d69157-1a56-4a78-a2f4-d7a134d7c3e9.mp4",
        "https://user-images.githubusercontent.com/28951144/229373718-86ce5e1d-d195-45d5-baa6-ef94041d0b90.mp4",
        "https://user-images.githubusercontent.com/28951144/229373716-76da0a4e-225a-44e4-9ee7-3e9006dbc3e3.mp4",
        "https://user-images.githubusercontent.com/28951144/229373695-22f88f13-d18f-4288-9bf1-c3e078d83722.mp4",
        "https://user-images.githubusercontent.com/28951144/229373709-603a7a89-2105-4e1b-a5a5-a6c3567c9a59.mp4",
    ]
    sample_media = []
    headers = {
        "User-agent": "mobile",
        "Authorization": "Bearer somelongtoken",
        "email": "myemail@test.com",
    }
    extras = {
        # audio_service configurables: title,artist,artUri,album,genre,displayTitle,displaySubtitle,displayDescription
        "artUri": "https://mp3.sunsaturn.com/crystaltunes.png", # background url for notification bar
    }
    for i in range(0, len(SONG_LIST)):
        if page.web:
            sample_media.append(ft.VideoMedia(SONG_LIST[i], extras=extras))
        else:
            sample_media.append(ft.VideoMedia(SONG_LIST[i], http_headers=headers, extras=extras))

    page.add(
        video := ft.Video(
            expand=True,
            playlist=sample_media,
            playlist_mode=ft.PlaylistMode.LOOP,
            fill_color=ft.Colors.BLUE_400,
            aspect_ratio=16/9,
            volume=100,
            autoplay=False,
            throttle=1000, # throttle on_position_changed this many ms(default 1000ms/1s)
            filter_quality=ft.FilterQuality.HIGH,
            muted=False,
            on_loaded=lambda e: print("Video loaded successfully!"),
            on_enter_fullscreen=lambda e: print("Video entered fullscreen!"),
            on_exit_fullscreen=lambda e: print("Video exited fullscreen!"),
            on_error=handle_error,
            on_track_changed=handle_track_changed,
            on_position_changed=handle_position_changed,

        ),
        ft.Row(
            wrap=True,
            alignment=ft.MainAxisAlignment.CENTER,
            controls=[
                ft.ElevatedButton("Play", on_click=handle_play),
                ft.ElevatedButton("Pause", on_click=handle_pause),
                ft.ElevatedButton("Play Or Pause", on_click=handle_play_or_pause),
                ft.ElevatedButton("Stop", on_click=handle_stop),
                ft.ElevatedButton("Next", on_click=handle_next),
                ft.ElevatedButton("Previous", on_click=handle_previous),
                ft.ElevatedButton("Seek s=10", on_click=handle_seek),
                ft.ElevatedButton("Jump to first Media", on_click=handle_jump),
                ft.ElevatedButton("Add Random Media", on_click=handle_add_media),
                ft.ElevatedButton("Remove Random Media", on_click=handle_remove_media),
            ],
        ),
        ft.Slider(
            min=0,
            value=100,
            max=100,
            label="Volume = {value}%",
            divisions=10,
            width=400,
            on_change=handle_volume_change,
        ),
        ft.Slider(
            min=1,
            value=1,
            max=3,
            label="PlaybackRate = {value}X",
            divisions=6,
            width=400,
            on_change=handle_playback_rate_change,
        ),
    )


ft.app(main)
```

syleishere commented:
Well if that's it for review, then Merry Christmas everyone, now everyone can code a Flet music/video player, install it on their phones, pop in some wireless bluetooth buds and head to gym in New Year playing your own app and music in background :)

ndonkoHenri commented:
> Commit: Flet Predictive Back Support Google TV

@syleishere please dont mix up stuffs.

syleishere commented:
Yeah I apologize, I normally wouldn't but Feodor asked me why I closed other PR, so figured he wanted it. Won't happen again.

ndonkoHenri commented:
Can you move that into a separate PR, please?

syleishere commented:
I am unsure how to do that, I can't seem to fork 2 copies of flet on github to do separate PR requests.

syleishere commented:
Alright that worked, created a branch called popscope, allowed me do another PR to same repo. Done...

syleishere commented:
Fixed race condition that happened when tracks switch and duration not updated yet, added buffer in seconds as someone may find it useful. Updating example with buffer, done with this PR, Merry Christmas.

```
import random
import flet as ft


def main(page: ft.Page):
    page.theme_mode = ft.ThemeMode.LIGHT
    page.title = "TheEthicalVideo"
    page.window.always_on_top = True
    page.spacing = 20
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER

    def handle_pause(e):
        video.pause()
        print("Video.pause()")

    def handle_play_or_pause(e):
        video.play_or_pause()
        print("Video.play_or_pause()")

    def handle_play(e):
        video.play()
        print("Video.play()")

    def handle_stop(e):
        video.stop()
        print("Video.stop()")

    def handle_next(e):
        video.next()
        print("Video.next()")

    def handle_previous(e):
        video.previous()
        print("Video.previous()")

    def handle_volume_change(e):
        video.volume = e.control.value
        page.update()
        print(f"Video.volume = {e.control.value}")

    def handle_playback_rate_change(e):
        video.playback_rate = e.control.value
        page.update()
        print(f"Video.playback_rate = {e.control.value}")

    def handle_seek(e):
        video.seek(10000)
        print(f"Video.seek(10000)")

    def handle_add_media(e):
        video.playlist_add(random.choice(sample_media))
        print(f"Video.playlist_add(random.choice(sample_media))")

    def handle_remove_media(e):
        r = random.randint(0, len(video.playlist) - 1)
        video.playlist_remove(r)
        print(f"Popped Item at index: {r} (position {r+1})")

    def handle_jump(e):
        print(f"Video.jump_to(0)")
        video.jump_to(0)

    def handle_track_changed(e):
        index = int(e.data)
        print(f"Track changed to index: {index}, Currently Playing {SONG_LIST[index]}")

    def handle_position_changed(e):
        print(f"Position is {e.position} seconds, duration is {e.duration} seconds, "
              f"buffer is {e.buffer} seconds, percent done is {e.percent}%")

    def handle_error(e):
        print(f"Video error: {e}")

    SONG_LIST = [
        "https://user-images.githubusercontent.com/28951144/229373720-14d69157-1a56-4a78-a2f4-d7a134d7c3e9.mp4",
        "https://user-images.githubusercontent.com/28951144/229373718-86ce5e1d-d195-45d5-baa6-ef94041d0b90.mp4",
        "https://user-images.githubusercontent.com/28951144/229373716-76da0a4e-225a-44e4-9ee7-3e9006dbc3e3.mp4",
        "https://user-images.githubusercontent.com/28951144/229373695-22f88f13-d18f-4288-9bf1-c3e078d83722.mp4",
        "https://user-images.githubusercontent.com/28951144/229373709-603a7a89-2105-4e1b-a5a5-a6c3567c9a59.mp4",
    ]
    sample_media = []
    headers = {
        "User-agent": "mobile",
        "Authorization": "Bearer somelongtoken",
        "email": "myemail@test.com",
    }
    extras = {
        # audio_service configurables: title,artist,artUri,album,genre,displayTitle,displaySubtitle,displayDescription
        "artUri": "https://mp3.sunsaturn.com/crystaltunes.png", # background url for notification bar
    }
    for i in range(0, len(SONG_LIST)):
        if page.web:
            sample_media.append(ft.VideoMedia(SONG_LIST[i], extras=extras))
        else:
            sample_media.append(ft.VideoMedia(SONG_LIST[i], http_headers=headers, extras=extras))

    page.add(
        video := ft.Video(
            expand=True,
            playlist=sample_media,
            playlist_mode=ft.PlaylistMode.LOOP,
            fill_color=ft.Colors.BLUE_400,
            aspect_ratio=16/9,
            volume=100,
            autoplay=False,
            throttle=1000, # throttle on_position_changed this many ms(default 1000ms/1s)
            filter_quality=ft.FilterQuality.HIGH,
            muted=False,
            on_loaded=lambda e: print("Video loaded successfully!"),
            on_enter_fullscreen=lambda e: print("Video entered fullscreen!"),
            on_exit_fullscreen=lambda e: print("Video exited fullscreen!"),
            on_error=handle_error,
            on_track_changed=handle_track_changed,
            on_position_changed=handle_position_changed,

        ),
        ft.Row(
            wrap=True,
            alignment=ft.MainAxisAlignment.CENTER,
            controls=[
                ft.ElevatedButton("Play", on_click=handle_play),
                ft.ElevatedButton("Pause", on_click=handle_pause),
                ft.ElevatedButton("Play Or Pause", on_click=handle_play_or_pause),
                ft.ElevatedButton("Stop", on_click=handle_stop),
                ft.ElevatedButton("Next", on_click=handle_next),
                ft.ElevatedButton("Previous", on_click=handle_previous),
                ft.ElevatedButton("Seek s=10", on_click=handle_seek),
                ft.ElevatedButton("Jump to first Media", on_click=handle_jump),
                ft.ElevatedButton("Add Random Media", on_click=handle_add_media),
                ft.ElevatedButton("Remove Random Media", on_click=handle_remove_media),
            ],
        ),
        ft.Slider(
            min=0,
            value=100,
            max=100,
            label="Volume = {value}%",
            divisions=10,
            width=400,
            on_change=handle_volume_change,
        ),
        ft.Slider(
            min=1,
            value=1,
            max=3,
            label="PlaybackRate = {value}X",
            divisions=6,
            width=400,
            on_change=handle_playback_rate_change,
        ),
    )


ft.app(main)
```

syleishere commented:
It adds ability to play audio or video in background, I normally just load mp3s and hide the video player till I want to play a video, but only way to have it keep playing on android with phone screen off or backgrounded, needs a foreground service with a notification to do that which audio_service provides.

You can just run it on an emulator for testing, I just hit the play button on android studio and select an emulator to use. When its installed on emulator, you can long press the app, go to permissions and just manually enable everything, every subsequent run of the app will run with full permissions.

syleishere commented:
![Google_TV](https://github.com/user-attachments/assets/a4bc9cde-55cc-41ad-aba5-2df85a3cb919)
<img width="1920" alt="macos_on_android_emulator" src="https://github.com/user-attachments/assets/e8bd5878-bec1-4778-92b7-2d0dc3e2d707" />
![Pixel8_real_device](https://github.com/user-attachments/assets/44a2f1a1-5739-493c-b2f8-1605d38c7730)
![windows_on_android_emulator](https://github.com/user-attachments/assets/c20fcc10-c303-4938-92e2-cf1918bdf0fc)

ndonkoHenri commented:
Flet-video has been moved to a seperate github repo: https://github.com/flet-dev/flet-video

syleishere commented:
I thought Feodor was going to look at this when he had time cause of the androidmanifest.xml changes etc.

litmount commented:
What is the status on this? This is exactly what I am looking for.

FeodorFitsner commented:
You need to implement this as a separate package, besides `flet-video` has been moved into a separate package too.

litmount commented:
> > You need to implement this as a separate package,
> 
> does this mean its part of the flet-video package code or does this have to be separately implemented on top of the existing flet-video implementation?

I've been wondering the status of this also. I made a request a long time ago about needing to have integration of audio player in the controls notification tray of android / iOS and as far as I know this has not happend

typedecker commented:
Thats rather unfortunate :/ I used the flet-audio widget, but the sound didn't last in the background for more than 6 minutes, and with another phone it lasted only 30 seconds, I also cannot make my own extension to solve this cuz flet build does not work for me for some reason, so right now I am just stuck with this. I thought of trying to use the flet-video widget for it instead, but I am not sure if it'll extend the time it keeps playing the audio for, have you tried using flet-video for the audio? how long did it work for you?

litmount commented:
Here is the discussion where I asked about these  https://github.com/flet-dev/flet/discussions/3291 
#3291 

Also see this for a possible work around https://github.com/flet-dev/flet-audio/issues/4#issuecomment-2820227496

typedecker commented:
@litmount I did actually come across the work around, but it seems like that extension has been made to keep the device's screen awake constantly right? i just want a way to play the audio even when the display is either off or if the app is in the background, have you tried the workaround? was it keeping the device screen awake?

litmount commented:
@typedecker No I didn't test that. I just came across it and thought it might be of help. I see what you mean though as far as just keeping the screen on.

typedecker commented:
@typedecker yeah 😅 I am hoping that once the extension build errors get fixed, I can prolly spin up an extension for flet that can do all the audio stuff.
